From c8e51a6aa2d9810e90e1fe766666a40e4a7264bc Mon Sep 17 00:00:00 2001
From: Nan Wu <wnan@google.com>
Date: Wed, 10 Sep 2025 21:31:17 +0000
Subject: [PATCH] RESTRICT AUTOMERGE: When an opaque window has alpha < 0.5,
 don't let it receive touches.

When a window is declared as opaque but the alpha is low, it's likely a fadeout animation or other visual change that should not affect touch handling.

Bug: 266433089
Test: inputflinger_tests
Flag: EXEMPT bug fix
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:9f65d0158f853ca4571a7af155969f0081d79a49
Merged-In: I7595c021e5f259ff77abd801a0d084212412c484
Change-Id: I7595c021e5f259ff77abd801a0d084212412c484
---
 services/inputflinger/dispatcher/InputDispatcher.cpp | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/services/inputflinger/dispatcher/InputDispatcher.cpp b/services/inputflinger/dispatcher/InputDispatcher.cpp
index 8419bbdd77..f7d9ba20f5 100644
--- a/services/inputflinger/dispatcher/InputDispatcher.cpp
+++ b/services/inputflinger/dispatcher/InputDispatcher.cpp
@@ -20,6 +20,7 @@
 #define LOG_NDEBUG 1
 
 #include <android-base/chrono_utils.h>
+#include <android-base/logging.h>
 #include <android-base/properties.h>
 #include <android-base/stringprintf.h>
 #include <android/os/IInputConstants.h>
@@ -2132,6 +2133,11 @@ InputEventInjectionResult InputDispatcher::findTouchedWindowTargetsLocked(
         for (const sp<WindowInfoHandle>& windowHandle : newTouchedWindows) {
             const WindowInfo& info = *windowHandle->getInfo();
 
+            if (info.touchOcclusionMode == TouchOcclusionMode::USE_OPACITY && info.alpha < 0.5f) {
+                LOG(INFO) << "Not sending motion to " << windowHandle->getName()
+                    << ", window opacity=" << info.alpha << " is below the threshold";
+                continue;
+            }
             // Skip spy window targets that are not valid for targeted injection.
             if (const auto err = verifyTargetedInjection(windowHandle, entry); err) {
                 continue;
-- 
2.52.0.rc1.455.g30608eb744-goog

